---
title: "phylosmith"
author: "Schuyler D. smith"
date: "`r Sys.Date()`"
output: 
  BiocStyle::html_document
vignette: >
    %\VignetteIndexEntry{phylosmith}
    %\VignetteEngine{knitr::rmarkdown}
---
# `phylosmith` 

[![Travis Build Status](https://travis-ci.org/schuyler-smith/phylosmith.svg?branch=master)](https://github.com/schuyler-smith/phylosmith)
[![status](http://joss.theoj.org/papers/4d4780ab16c487764ebe108fa6bcdc2c/status.svg)](http://joss.theoj.org/papers/4d4780ab16c487764ebe108fa6bcdc2c)

[![DOI](https://zenodo.org/badge/159598780.svg)](https://zenodo.org/badge/latestdoi/159598780)

A supplementary package to build on the [phyloseq](https://github.com/joey711/phyloseq) package. Phyloseq objects are a great data-standard for microbiome and gene-expression data, this package is aimed to provied easy data-wrangling and visualization.

# Installation
## Requirements
### Linux Systems 
For some Linux systems you may need to install the following two programs through your terminal.

Ubuntu example:
```bash
sudo apt-get install libmysqlclient-dev libgdal-dev libudunits2-dev
```
These programs are required by some dependencies and may not come in your default OS distribution.

### Windows Systems 

if you are working on `WINDOWS` you likely need to install the CRAN program <a href="https://cran.r-project.org/bin/windows/Rtools/" target="_blank" >Rtools</a>. When prompted, select `add rtools to system PATH`.

### R

phylosmith depends on the usage of the [phyloseq package](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0061217) released by [Dr. Paul McMurdie](https://github.com/joey711). The package is maintained on BioConductor, and can be installed through R using the following commands:

```
if(!requireNamespace("BiocManager", quietly = TRUE)){
  install.packages("BiocManager")
} 
BiocManager::install("phyloseq")
```

Additionally, the package imports a number of other packages to use their advanced functions. These packages may install with the phylosmith installation, but it is always best to install independently.

```
install.packages(c("devtools", "RcppEigen", "RcppParallel", "Rtsne", "ggforce", "units"))
```

## phylosmith

The package is hosted on Github, and can be installed through R with:
```
remotes::install_github('schuyler-smith/phylosmith')
library(phylosmith)
```


# Functions

## [Wrangling](https://schuyler-smith.github.io/phylosmith/wrangling.html)

Call | Description
-------------------- | ------------------------------------------------------------
[conglomerate_samples](https://schuyler-smith.github.io/phylosmith/wrangling.html#conglomerate_samples) | combines samples based on common factor within sample_data
[conglomerate_taxa](https://schuyler-smith.github.io/phylosmith/wrangling.html#conglomerate_taxa) | combines taxa that have same classification
[melt_phyloseq](https://schuyler-smith.github.io/phylosmith/wrangling.html#melt_phyloseq) | melts a phyloseq object into a data.table
[merge_treatments](https://schuyler-smith.github.io/phylosmith/wrangling.html#merge_treatments) | combines multiple columns in meta-data into a new column
[relative_abundance](https://schuyler-smith.github.io/phylosmith/wrangling.html#relative_abundance) | transform abundance data to relative abundance
[set_sample_order](https://schuyler-smith.github.io/phylosmith/wrangling.html#set_sample_order) | sets the order of the samples of a phyloseq object
[set_treatment_levels](https://schuyler-smith.github.io/phylosmith/wrangling.html#set_treatment_levels) | sets the order of the factors in a sample_data column
[taxa_filter](https://schuyler-smith.github.io/phylosmith/wrangling.html#taxa_filter) | filter taxa by proportion of samples seen in

## [Analytics](https://schuyler-smith.github.io/phylosmith/analytics.html)
Call | Description
-------------------- | ------------------------------------------------------------
[common_taxa](https://schuyler-smith.github.io/phylosmith/analytics.html#common_taxa) | find taxa common to each treatment
[taxa_core](https://schuyler-smith.github.io/phylosmith/analytics.html#taxa_core) | filter taxa by proportion of samples and relative abundance
[taxa_proportions](https://schuyler-smith.github.io/phylosmith/analytics.html#taxa_proportions) | computes the proportion of a taxa classification
[unique_taxa](https://schuyler-smith.github.io/phylosmith/analytics.html#unique_taxa) | find taxa unique to each treatment

## [Graphs](https://schuyler-smith.github.io/phylosmith/graphs.html)
### Abundance
Call | Description
-------------------- | ------------------------------------------------------------
[abundance_heatmap](https://schuyler-smith.github.io/phylosmith/graphs.html#abundance_heatmap) | create a ggplot object of the heatmaps of the abundance table
[abundance_lines](https://schuyler-smith.github.io/phylosmith/graphs.html#abundance_lines) | create a ggplot object of the abundance data as a line graph
[phylogeny_profile](https://schuyler-smith.github.io/phylosmith/graphs.html#phylogeny_profile) | create a ggplot barplot object of the compositons of each sample at a taxonomic level
[taxa_abundance_bars](https://schuyler-smith.github.io/phylosmith/graphs.html#taxa_abundance_bars) | create a ggplot object of the abundance of taxa in each sample
[taxa_core_graph](https://schuyler-smith.github.io/phylosmith/graphs.html#taxa_core_graph) | create a ggplot object of the core taxa over a range of parameters
[variable_correlation_heatmap](https://schuyler-smith.github.io/phylosmith/graphs.html#variable_correlation_heatmap)  | create a ggplot heatmatp of the correlation of numerical variables with taxa

### Diversity
Call | Description
-------------------- | ------------------------------------------------------------
[alpha_diversity_graph](https://schuyler-smith.github.io/phylosmith/graphs.html#alpha_diversity_graph) | create a ggplot-object box-plot of the alpha-diversity from a phyloseq-object.
[nmds_phyloseq](https://schuyler-smith.github.io/phylosmith/graphs.html#nmds_phyloseq)  | create a ggplot object of the NMDS from a phyloseq object
[pcoa_phyloseq](https://schuyler-smith.github.io/phylosmith/graphs.html#pcoa_phyloseq)  | create a ggplot object of the PCoA from a phyloseq object
[tsne_phyloseq](https://schuyler-smith.github.io/phylosmith/graphs.html#tsne_phyloseq)  | create a ggplot object of the t-SNE from a phyloseq object

## [Networks](https://schuyler-smith.github.io/phylosmith/networks.html)
Call | Description
-------------------- | ------------------------------------------------------------
[co_occurrence_network](https://schuyler-smith.github.io/phylosmith/networks.html#co_occurrence_network) | creates a network of the co-occurrence of taxa
[network_layout_ps](https://schuyler-smith.github.io/phylosmith/networks.html#network_layout_ps) | creates a layout object for a network
[variable_correlation_network](https://schuyler-smith.github.io/phylosmith/networks.html#variable_correlation_network) | creates a network of the correlation of taxa and sample variables

## [Calculations](https://schuyler-smith.github.io/phylosmith/calculations.html)
Call | Description
-------------------- | ------------------------------------------------------------
[co_occurrence](https://schuyler-smith.github.io/phylosmith/calculations.html#co_occurrence) | calculate co-occurrence between taxa
[permute_rho](https://schuyler-smith.github.io/phylosmith/calculations.html#permute_rho) | runs permutations of the otu_table to calculate a significant $\rho$ value
[histogram_permuted_rhos](https://schuyler-smith.github.io/phylosmith/calculations.html#histogram_permuted_rhos) | Create a ggplot object of the distribution of rho values.
[quantile_permuted_rhos](https://schuyler-smith.github.io/phylosmith/calculations.html#quantile_permuted_rhos) | calculate quantiles for the permuted rho values from the Spearman-rank co-occurrence
[variable_correlation](https://schuyler-smith.github.io/phylosmith/calculations.html#variable_correlation) | calculate the correlation of numerical variables with taxa abundances

# Datasets

Originally I had created 2 mock phyloseq objects (`mock_phyloseq` and 
`mock_phyloseq2`) that had no real-world data but served to show simple 
examples of how the functions worked. 

Then I decided that I should include a real example of microbiome data 
(`soil_column`) because it's always nice to see real examples. 
`soil_column` is a <a href="https://www.frontiersin.org/
articles/10.3389/fmicb.2018.03197/full" target="_blank">published dataset</a> 
from my lab-group. The data is from an experiment 
where they looked at the microbial composition of farmland soil before and 
after manure application, over time, using 16S-sequencing.

<br>
<br>
