<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Schuyler D. Smith" />

<meta name="date" content="2023-03-26" />

<title>Calculations</title>

<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cerulean.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="src/styles.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
}

</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">phylosmith</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
    Home
  </a>
</li>
<li>
  <a href="wrangling.html">
    <span class="fa fa-file"></span>
     
    Wrangling
  </a>
</li>
<li>
  <a href="analytics.html">
    <span class="fa fa-chart-pie"></span>
     
    Analytics
  </a>
</li>
<li>
  <a href="graphics.html">
    <span class="fa fa-bar-chart"></span>
     
    Graphics
  </a>
</li>
<li>
  <a href="networks.html">
    <span class="fa fa-project-diagram"></span>
     
    Networks
  </a>
</li>
<li>
  <a href="calculations.html">
    <span class="fa fa-calculator"></span>
     
    Calculations
  </a>
</li>
<li>
  <a href="https://github.com/schuyler-smith/phylosmith/">
    <span class="fa fa-code"></span>
     
    Source Code
  </a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="http://github.com/schuyler-smith">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="http://germslab.org/">
    <span class="fa fa-flask"></span>
     
  </a>
</li>
<li>
  <a href="https://www.bcb.iastate.edu/people/schuyler-smith">
    <span class="fa fa-university"></span>
     
  </a>
</li>
<li>
  <a href="mailto:sdsmith@iastate.edu">
    <span class="fa fa-envelope fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Calculations</h1>
<h4 class="author"><a href="https://schuyler-smith.github.io/" target="_blank" >Schuyler D. Smith</a></h4>
<h4 class="date">2023-03-26</h4>

</div>


<p>Examples used in this vignette will use the <code>GlobalPatterns</code> dataset from <code>phyloseq</code>.</p>
<pre class="r"><code>library(phyloseq)
data(GlobalPatterns)</code></pre>
<p><br></p>
<div id="co_occurrence" class="section level2">
<h2>co_occurrence</h2>
<p><a name="co_occurrence"></a></p>
<p>This function is an implementation of the pair-wise Spearman rank correlation that you can call in <code>R</code>. The results should be identical. <a href="https://github.com/germs-lab/FastCoOccur" target="_blank" >FastCoOccur</a> was originally written by <a href="https://github.com/metajinomics" target="_blank" >Jin Choi</a> in C++ to be called from the command-line. I rewrote the routine, implementing the <a href="http://www.rcpp.org/" target="_blank" >Rcpp</a> API, so that it could be called within R, and with a phyloseq-object as the input.</p>
<p>Because of how I implemented the multithreading, there is a large diminishing return to the number of cores used. I found the optimal to be around 4 CPUs, depending on the size of your dataset.</p>
<p><br> <strong><em>Usage</em></strong></p>
<pre class="r"><code>co_occurrence(phyloseq_obj, treatment = NULL, subset = NULL, rho = 0, p = 0.05, cores = 0)</code></pre>
<p><br> <strong><em>Arguments</em></strong></p>
<table>
<colgroup>
<col width="25%" />
<col width="75%" />
</colgroup>
<thead>
<tr class="header">
<th>Call</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>phyloseq_obj</code></td>
<td>A phyloseq-class object.</td>
</tr>
<tr class="even">
<td><code>treatment</code></td>
<td>Column name as a <code>string</code>, or <code>vector</code> of, in the <code>sample_data</code>.</td>
</tr>
<tr class="odd">
<td><code>p</code></td>
<td>The p-value cutoff. All returned co-occurrences will have a p-value less than or equal to <code>p</code>.</td>
</tr>
<tr class="even">
<td><code>cores</code></td>
<td>Number of CPU cores to use for the pair-wise permutations. Default (0) uses max cores available. Parallelization not available for systems running MacOS without openMP</td>
</tr>
</tbody>
</table>
<p><strong><em>Examples</em></strong></p>
<pre class="r"><code>filtered_obj &lt;- conglomerate_taxa(GlobalPatterns, &quot;Phylum&quot;)
co_occurrence(filtered_obj, &quot;SampleType&quot;, rho = 0.8, p = 0.05)</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["Treatment"],"name":[1],"type":["chr"],"align":["left"]},{"label":["X"],"name":[2],"type":["chr"],"align":["left"]},{"label":["Y"],"name":[3],"type":["chr"],"align":["left"]},{"label":["rho"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["p"],"name":[5],"type":["dbl"],"align":["right"]}],"data":[{"1":"Soil","2":"SBR1093","3":"ZB2","4":"1","5":"9.486374e-09"},{"1":"Soil","2":"SBR1093","3":"Synergistetes","4":"-1","5":"9.486374e-09"},{"1":"Soil","2":"GOUTA4","3":"OP8","4":"-1","5":"9.486374e-09"},{"1":"Soil","2":"GOUTA4","3":"GN04","4":"1","5":"9.486374e-09"},{"1":"Soil","2":"GOUTA4","3":"SR1","4":"1","5":"9.486374e-09"},{"1":"Soil","2":"OP8","3":"GN04","4":"-1","5":"9.486374e-09"},{"1":"Soil","2":"OP8","3":"SR1","4":"-1","5":"9.486374e-09"},{"1":"Soil","2":"Hyd24-12","3":"OP11","4":"1","5":"9.486374e-09"},{"1":"Soil","2":"GN04","3":"SR1","4":"1","5":"9.486374e-09"},{"1":"Soil","2":"ZB2","3":"Synergistetes","4":"-1","5":"9.486374e-09"},{"1":"Feces","2":"Euryarchaeota","3":"Chlorobi","4":"1","5":"2.220446e-16"},{"1":"Feces","2":"Euryarchaeota","3":"Chloroflexi","4":"1","5":"2.220446e-16"},{"1":"Feces","2":"Actinobacteria","3":"Fusobacteria","4":"-1","5":"2.220446e-16"},{"1":"Feces","2":"Actinobacteria","3":"Bacteroidetes","4":"-1","5":"2.220446e-16"},{"1":"Feces","2":"Actinobacteria","3":"Lentisphaerae","4":"-1","5":"2.220446e-16"},{"1":"Feces","2":"Actinobacteria","3":"Firmicutes","4":"1","5":"2.220446e-16"},{"1":"Feces","2":"Proteobacteria","3":"Planctomycetes","4":"-1","5":"2.220446e-16"},{"1":"Feces","2":"Fusobacteria","3":"Bacteroidetes","4":"1","5":"2.220446e-16"},{"1":"Feces","2":"Fusobacteria","3":"Lentisphaerae","4":"1","5":"2.220446e-16"},{"1":"Feces","2":"Fusobacteria","3":"Firmicutes","4":"-1","5":"2.220446e-16"},{"1":"Feces","2":"Chlorobi","3":"Chloroflexi","4":"1","5":"2.220446e-16"},{"1":"Feces","2":"Bacteroidetes","3":"Lentisphaerae","4":"1","5":"2.220446e-16"},{"1":"Feces","2":"Bacteroidetes","3":"Firmicutes","4":"-1","5":"2.220446e-16"},{"1":"Feces","2":"SAR406","3":"OP8","4":"1","5":"2.220446e-16"},{"1":"Feces","2":"SAR406","3":"Synergistetes","4":"-1","5":"2.220446e-16"},{"1":"Feces","2":"Gemmatimonadetes","3":"SPAM","4":"1","5":"2.220446e-16"},{"1":"Feces","2":"OP8","3":"Synergistetes","4":"-1","5":"2.220446e-16"},{"1":"Feces","2":"Lentisphaerae","3":"Firmicutes","4":"-1","5":"2.220446e-16"},{"1":"Skin","2":"Euryarchaeota","3":"SBR1093","4":"1","5":"9.486374e-09"},{"1":"Skin","2":"Euryarchaeota","3":"OP8","4":"1","5":"9.486374e-09"},{"1":"Skin","2":"Euryarchaeota","3":"CCM11b","4":"1","5":"9.486374e-09"},{"1":"Skin","2":"SBR1093","3":"OP8","4":"1","5":"9.486374e-09"},{"1":"Skin","2":"SBR1093","3":"CCM11b","4":"1","5":"9.486374e-09"},{"1":"Skin","2":"TM6","3":"Unclassified Bacteria","4":"-1","5":"9.486374e-09"},{"1":"Skin","2":"TM6","3":"WS3","4":"-1","5":"9.486374e-09"},{"1":"Skin","2":"TM6","3":"Lentisphaerae","4":"1","5":"9.486374e-09"},{"1":"Skin","2":"OP8","3":"CCM11b","4":"1","5":"9.486374e-09"},{"1":"Skin","2":"Elusimicrobia","3":"GN04","4":"-1","5":"9.486374e-09"},{"1":"Skin","2":"Elusimicrobia","3":"AD3","4":"1","5":"9.486374e-09"},{"1":"Skin","2":"Unclassified Bacteria","3":"WS3","4":"1","5":"9.486374e-09"},{"1":"Skin","2":"Unclassified Bacteria","3":"Lentisphaerae","4":"-1","5":"9.486374e-09"},{"1":"Skin","2":"WS3","3":"Lentisphaerae","4":"-1","5":"9.486374e-09"},{"1":"Skin","2":"GN04","3":"AD3","4":"-1","5":"9.486374e-09"},{"1":"Tongue","2":"Thermi","3":"GAL15","4":"-1","5":"9.486374e-09"},{"1":"Freshwater","2":"MVP-15","3":"KSB1","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"MVP-15","3":"TM6","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"MVP-15","3":"NC10","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"MVP-15","3":"NKB19","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"MVP-15","3":"BRC1","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"MVP-15","3":"Hyd24-12","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"MVP-15","3":"OP9","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"MVP-15","3":"WS1","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"ZB3","3":"GN06","4":"-1","5":"9.486374e-09"},{"1":"Freshwater","2":"ZB3","3":"SPAM","4":"-1","5":"9.486374e-09"},{"1":"Freshwater","2":"ZB3","3":"WPS-2","4":"-1","5":"9.486374e-09"},{"1":"Freshwater","2":"ZB3","3":"SC3","4":"-1","5":"9.486374e-09"},{"1":"Freshwater","2":"KSB1","3":"TM6","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"KSB1","3":"NC10","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"KSB1","3":"NKB19","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"KSB1","3":"BRC1","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"KSB1","3":"Hyd24-12","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"KSB1","3":"OP9","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"KSB1","3":"WS1","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"Thermi","3":"SC4","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"Thermi","3":"SR1","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"GN06","3":"SPAM","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"GN06","3":"WPS-2","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"GN06","3":"SC3","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"TM6","3":"NC10","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"TM6","3":"NKB19","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"TM6","3":"BRC1","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"TM6","3":"Hyd24-12","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"TM6","3":"OP9","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"TM6","3":"WS1","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"NC10","3":"NKB19","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"NC10","3":"BRC1","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"NC10","3":"Hyd24-12","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"NC10","3":"OP9","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"NC10","3":"WS1","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"SPAM","3":"WPS-2","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"SPAM","3":"SC3","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"NKB19","3":"BRC1","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"NKB19","3":"Hyd24-12","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"NKB19","3":"OP9","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"NKB19","3":"WS1","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"BRC1","3":"Hyd24-12","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"BRC1","3":"OP9","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"BRC1","3":"WS1","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"Hyd24-12","3":"OP9","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"Hyd24-12","3":"WS1","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"OP9","3":"WS1","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"WPS-2","3":"SC3","4":"1","5":"9.486374e-09"},{"1":"Freshwater","2":"SC4","3":"SR1","4":"1","5":"9.486374e-09"},{"1":"Freshwater (creek)","2":"GOUTA4","3":"NC10","4":"1","5":"9.486374e-09"},{"1":"Freshwater (creek)","2":"GOUTA4","3":"LD1","4":"1","5":"9.486374e-09"},{"1":"Freshwater (creek)","2":"GOUTA4","3":"GN02","4":"-1","5":"9.486374e-09"},{"1":"Freshwater (creek)","2":"GOUTA4","3":"WS1","4":"1","5":"9.486374e-09"},{"1":"Freshwater (creek)","2":"GOUTA4","3":"WS2","4":"1","5":"9.486374e-09"},{"1":"Freshwater (creek)","2":"TG3","3":"GN06","4":"1","5":"9.486374e-09"},{"1":"Freshwater (creek)","2":"TG3","3":"WPS-2","4":"1","5":"9.486374e-09"},{"1":"Freshwater (creek)","2":"GN06","3":"WPS-2","4":"1","5":"9.486374e-09"},{"1":"Freshwater (creek)","2":"AC1","3":"CCM11b","4":"1","5":"9.486374e-09"},{"1":"Freshwater (creek)","2":"AC1","3":"SM2F11","4":"1","5":"9.486374e-09"},{"1":"Freshwater (creek)","2":"NC10","3":"LD1","4":"1","5":"9.486374e-09"},{"1":"Freshwater (creek)","2":"NC10","3":"GN02","4":"-1","5":"9.486374e-09"},{"1":"Freshwater (creek)","2":"NC10","3":"WS1","4":"1","5":"9.486374e-09"},{"1":"Freshwater (creek)","2":"NC10","3":"WS2","4":"1","5":"9.486374e-09"},{"1":"Freshwater (creek)","2":"CCM11b","3":"SM2F11","4":"1","5":"9.486374e-09"},{"1":"Freshwater (creek)","2":"LD1","3":"GN02","4":"-1","5":"9.486374e-09"},{"1":"Freshwater (creek)","2":"LD1","3":"WS1","4":"1","5":"9.486374e-09"},{"1":"Freshwater (creek)","2":"LD1","3":"WS2","4":"1","5":"9.486374e-09"},{"1":"Freshwater (creek)","2":"GN02","3":"WS1","4":"-1","5":"9.486374e-09"},{"1":"Freshwater (creek)","2":"GN02","3":"WS2","4":"-1","5":"9.486374e-09"},{"1":"Freshwater (creek)","2":"WS1","3":"WS2","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"Thermi","3":"CCM11b","4":"-1","5":"9.486374e-09"},{"1":"Ocean","2":"Thermi","3":"BRC1","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"Thermi","3":"Chlamydiae","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"Fibrobacteres","3":"Nitrospirae","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"Fibrobacteres","3":"NKB19","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"Fibrobacteres","3":"GN04","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"Fibrobacteres","3":"Lentisphaerae","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"Fibrobacteres","3":"OP3","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"Fibrobacteres","3":"OP9","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"Fibrobacteres","3":"GN02","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"Fibrobacteres","3":"WS1","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"Fibrobacteres","3":"SR1","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"TM6","3":"GAL15","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"CCM11b","3":"BRC1","4":"-1","5":"9.486374e-09"},{"1":"Ocean","2":"CCM11b","3":"Chlamydiae","4":"-1","5":"9.486374e-09"},{"1":"Ocean","2":"Nitrospirae","3":"NKB19","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"Nitrospirae","3":"GN04","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"Nitrospirae","3":"Lentisphaerae","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"Nitrospirae","3":"OP3","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"Nitrospirae","3":"OP9","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"Nitrospirae","3":"GN02","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"Nitrospirae","3":"WS1","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"Nitrospirae","3":"SR1","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"NKB19","3":"GN04","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"NKB19","3":"Lentisphaerae","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"NKB19","3":"OP3","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"NKB19","3":"OP9","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"NKB19","3":"GN02","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"NKB19","3":"WS1","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"NKB19","3":"SR1","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"BRC1","3":"Chlamydiae","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"GN04","3":"Lentisphaerae","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"GN04","3":"OP3","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"GN04","3":"OP9","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"GN04","3":"GN02","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"GN04","3":"WS1","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"GN04","3":"SR1","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"Lentisphaerae","3":"OP3","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"Lentisphaerae","3":"OP9","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"Lentisphaerae","3":"GN02","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"Lentisphaerae","3":"WS1","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"Lentisphaerae","3":"SR1","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"OP3","3":"OP9","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"OP3","3":"GN02","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"OP3","3":"WS1","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"OP3","3":"SR1","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"OP9","3":"GN02","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"OP9","3":"WS1","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"OP9","3":"SR1","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"GN02","3":"WS1","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"GN02","3":"SR1","4":"1","5":"9.486374e-09"},{"1":"Ocean","2":"WS1","3":"SR1","4":"1","5":"9.486374e-09"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<p><br></p>
<hr />
<p><br></p>
</div>
<div id="permute_rho" class="section level2">
<h2>permute_rho</h2>
<p><a name="permute_rho"></a></p>
<p>Used to determine a cutoff for which <span class="math inline">\(\rho\)</span> vlaues are not likely to have occured by random chance. This will create permutations of all combinations of abundances and calculate the pair-wise Spearman rank co-occurence for each. The return will be a distribution of <span class="math inline">\(\rho\)</span> values that can be used to determing the significance.</p>
<p>The permutation returns a large table of rho-values, so RAM becomes a bottle-neck resource. To deal with this, the results are merged into a single frequency table, which saves an enormous amount of memory. Unfortunately, this doesn’t allow for simple and intuitive multithreading. Because of that, there is a large diminishing return to the number of cores used. I found the optimal to be around 4 CPUs, depending on the size of your dataset.</p>
<p><br> <strong><em>Usage</em></strong></p>
<pre class="r"><code>permute_rho(phyloseq_obj, treatment, replicate_samples = &#39;independent&#39;, permutations = 100, cores = 0)</code></pre>
<p><br> <strong><em>Arguments</em></strong></p>
<table>
<colgroup>
<col width="25%" />
<col width="75%" />
</colgroup>
<thead>
<tr class="header">
<th>Call</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>phyloseq_obj</code></td>
<td>A phyloseq-class object.</td>
</tr>
<tr class="even">
<td><code>treatment</code></td>
<td>Column name as a <code>string</code>, or <code>vector</code> of, in the <code>sample_data</code>.</td>
</tr>
<tr class="odd">
<td><code>replicate_samples</code></td>
<td>Column name as a string or number in the sample_data that indicates which samples are non-independent of each other.</td>
</tr>
<tr class="even">
<td><code>permutations</code></td>
<td>Number of iterations to compute.</td>
</tr>
<tr class="odd">
<td><code>cores</code></td>
<td>Number of CPU cores to use for the pair-wise permutations. Default (0) uses max cores available. Parallelization not available for systems running MacOS without openMP configuration.</td>
</tr>
</tbody>
</table>
<p><strong><em>Examples</em></strong></p>
<pre class="r"><code>filtered_obj &lt;- conglomerate_taxa(GlobalPatterns, &quot;Phylum&quot;)</code></pre>
<pre><code>## Warning in conglomerate_taxa(GlobalPatterns, &quot;Phylum&quot;): trees cannot be
## preserved after taxa conglomeration</code></pre>
<pre class="r"><code>permute_rho(filtered_obj, &quot;SampleType&quot;, replicate_samples = &quot;SampleType&quot;, permutations = 5)</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["Treatment"],"name":[1],"type":["chr"],"align":["left"]},{"label":["rho"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["Count"],"name":[3],"type":["int"],"align":["right"]}],"data":[{"1":"Feces","2":"-1.00","3":"5"},{"1":"Feces","2":"-0.95","3":"77"},{"1":"Feces","2":"-0.94","3":"6"},{"1":"Feces","2":"-0.89","3":"62"},{"1":"Feces","2":"-0.82","3":"10"},{"1":"Feces","2":"-0.80","3":"20"},{"1":"Feces","2":"-0.77","3":"229"},{"1":"Feces","2":"-0.74","3":"76"},{"1":"Feces","2":"-0.63","3":"30"},{"1":"Feces","2":"-0.60","3":"4"},{"1":"Feces","2":"-0.58","3":"5"},{"1":"Feces","2":"-0.54","3":"138"},{"1":"Feces","2":"-0.50","3":"2"},{"1":"Feces","2":"-0.45","3":"5"},{"1":"Feces","2":"-0.40","3":"26"},{"1":"Feces","2":"-0.39","3":"21"},{"1":"Feces","2":"-0.33","3":"600"},{"1":"Feces","2":"-0.32","3":"28"},{"1":"Feces","2":"-0.26","3":"322"},{"1":"Feces","2":"-0.24","3":"2"},{"1":"Feces","2":"-0.21","3":"36"},{"1":"Feces","2":"-0.20","3":"15"},{"1":"Feces","2":"-0.11","3":"7"},{"1":"Feces","2":"-0.06","3":"34"},{"1":"Feces","2":"0.00","3":"6069"},{"1":"Feces","2":"0.11","3":"6"},{"1":"Feces","2":"0.20","3":"6"},{"1":"Feces","2":"0.21","3":"40"},{"1":"Feces","2":"0.24","3":"1"},{"1":"Feces","2":"0.26","3":"186"},{"1":"Feces","2":"0.27","3":"712"},{"1":"Feces","2":"0.32","3":"27"},{"1":"Feces","2":"0.33","3":"9"},{"1":"Feces","2":"0.40","3":"17"},{"1":"Feces","2":"0.45","3":"5"},{"1":"Feces","2":"0.50","3":"18"},{"1":"Feces","2":"0.58","3":"100"},{"1":"Feces","2":"0.60","3":"7"},{"1":"Feces","2":"0.63","3":"41"},{"1":"Feces","2":"0.71","3":"1"},{"1":"Feces","2":"0.74","3":"54"},{"1":"Feces","2":"0.77","3":"264"},{"1":"Feces","2":"0.78","3":"127"},{"1":"Feces","2":"0.80","3":"18"},{"1":"Feces","2":"0.82","3":"685"},{"1":"Feces","2":"0.83","3":"4"},{"1":"Feces","2":"0.89","3":"6"},{"1":"Feces","2":"0.94","3":"30"},{"1":"Feces","2":"0.95","3":"56"},{"1":"Feces","2":"1.00","3":"806"},{"1":"Freshwater","2":"NaN","3":"162"},{"1":"Freshwater","2":"-1.00","3":"876"},{"1":"Freshwater","2":"0.00","3":"8914"},{"1":"Freshwater","2":"1.00","3":"1103"},{"1":"Freshwater (creek)","2":"-1.00","3":"354"},{"1":"Freshwater (creek)","2":"-0.87","3":"661"},{"1":"Freshwater (creek)","2":"-0.50","3":"624"},{"1":"Freshwater (creek)","2":"0.00","3":"7315"},{"1":"Freshwater (creek)","2":"0.50","3":"537"},{"1":"Freshwater (creek)","2":"0.87","3":"849"},{"1":"Freshwater (creek)","2":"1.00","3":"715"},{"1":"Mock","2":"-1.00","3":"29"},{"1":"Mock","2":"-0.87","3":"222"},{"1":"Mock","2":"-0.50","3":"168"},{"1":"Mock","2":"0.00","3":"10134"},{"1":"Mock","2":"0.50","3":"102"},{"1":"Mock","2":"0.87","3":"206"},{"1":"Mock","2":"1.00","3":"194"},{"1":"Ocean","2":"-1.00","3":"146"},{"1":"Ocean","2":"-0.87","3":"793"},{"1":"Ocean","2":"-0.50","3":"764"},{"1":"Ocean","2":"0.00","3":"6981"},{"1":"Ocean","2":"0.50","3":"295"},{"1":"Ocean","2":"0.87","3":"1022"},{"1":"Ocean","2":"1.00","3":"1054"},{"1":"Sediment (estuary)","2":"NaN","3":"19"},{"1":"Sediment (estuary)","2":"-1.00","3":"322"},{"1":"Sediment (estuary)","2":"-0.87","3":"515"},{"1":"Sediment (estuary)","2":"-0.50","3":"619"},{"1":"Sediment (estuary)","2":"0.00","3":"7455"},{"1":"Sediment (estuary)","2":"0.50","3":"740"},{"1":"Sediment (estuary)","2":"0.87","3":"750"},{"1":"Sediment (estuary)","2":"1.00","3":"635"},{"1":"Skin","2":"-1.00","3":"138"},{"1":"Skin","2":"-0.87","3":"257"},{"1":"Skin","2":"-0.50","3":"258"},{"1":"Skin","2":"0.00","3":"9574"},{"1":"Skin","2":"0.50","3":"236"},{"1":"Skin","2":"0.87","3":"286"},{"1":"Skin","2":"1.00","3":"306"},{"1":"Soil","2":"-1.00","3":"1460"},{"1":"Soil","2":"0.00","3":"8134"},{"1":"Soil","2":"1.00","3":"1461"},{"1":"Tongue","2":"NaN","3":"18"},{"1":"Tongue","2":"-1.00","3":"772"},{"1":"Tongue","2":"0.00","3":"9386"},{"1":"Tongue","2":"1.00","3":"879"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<p><br></p>
<hr />
<p><br></p>
</div>
<div id="histogram_permuted_rhos" class="section level2">
<h2>histogram_permuted_rhos</h2>
<p><a name="histogram_permuted_rhos"></a></p>
<p>Creates a ggplot object of the histogram of the rho values by treatment. This is a visualization tool to explain how to find cutoffs, but will also calculate the cutoffs.</p>
<p><br> <strong><em>Usage</em></strong></p>
<pre class="r"><code>histogram_permuted_rhos(permuted_rhos, p = 0.05, x_breaks = 0.25, colors = &#39;default&#39;)</code></pre>
<p><br> <strong><em>Arguments</em></strong></p>
<table>
<colgroup>
<col width="25%" />
<col width="75%" />
</colgroup>
<thead>
<tr class="header">
<th>Call</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>permuted_rhos</code></td>
<td>A <code>data.table</code> output from <code>permute_rho</code>.</td>
</tr>
<tr class="even">
<td><code>p</code></td>
<td>The significance threshold for setting cutoffs.</td>
</tr>
<tr class="odd">
<td><code>x_breaks</code></td>
<td>What intervals to set the ticks on the x-axis.</td>
</tr>
<tr class="even">
<td><code>colors</code></td>
<td>Name of a color set from the <a href="https://cran.r-project.org/web/packages/RColorBrewer/RColorBrewer.pdf" target="_blank" ><code>RColorBrewer</code></a> package or a vector palete of R-accepted colors.</td>
</tr>
</tbody>
</table>
<p><strong><em>Examples</em></strong></p>
<pre class="r"><code>filtered_obj &lt;- conglomerate_taxa(GlobalPatterns, &quot;Phylum&quot;)
permuted_rhos &lt;- permute_rho(filtered_obj, &quot;SampleType&quot;, replicate_samples = &quot;SampleType&quot;, permutations = 5)
histogram_permuted_rhos(permuted_rhos, p = 0.05)</code></pre>
<p><img src="calculations_files/figure-html/histogram_permuted_rhos-1.png" width="1536" /></p>
<p><br></p>
<hr />
<p><br></p>
</div>
<div id="quantile_permuted_rhos" class="section level2">
<h2>quantile_permuted_rhos</h2>
<p><a name="quantile_permuted_rhos"></a></p>
<p>Calculate quantiles for the permuted rho values from the Spearman-rank co-occurrence. Can be done by treatments or for the overall data. This looks at both tails of the distribution, so it will show the p/2 cutoffs on both ends.</p>
<p><br> <strong><em>Usage</em></strong></p>
<pre class="r"><code>quantile_permuted_rhos(permuted_rhos, p = 0.05, by_treatment = TRUE)</code></pre>
<p><br> <strong><em>Arguments</em></strong></p>
<table>
<colgroup>
<col width="25%" />
<col width="75%" />
</colgroup>
<thead>
<tr class="header">
<th>Call</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>permuted_rhos</code></td>
<td>A <code>data.table</code> output from <code>permute_rho</code>.</td>
</tr>
<tr class="even">
<td><code>p</code></td>
<td>The significance threshold for setting cutoffs.</td>
</tr>
<tr class="odd">
<td><code>by_treatment</code></td>
<td>Whether to find the rho cutoffs for each treatment individually or for the entire experiment. Suggested to do by treatment first, to see if there is any treatments that are outliers.</td>
</tr>
</tbody>
</table>
<p><strong><em>Examples</em></strong></p>
<pre class="r"><code>filtered_obj &lt;- conglomerate_taxa(GlobalPatterns, &quot;Phylum&quot;)
permuted_rhos &lt;- permute_rho(filtered_obj, &quot;SampleType&quot;, replicate_samples = &quot;SampleType&quot;, permutations = 5)
quantile_permuted_rhos(permuted_rhos, p = 0.05)</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["Treatment"],"name":[1],"type":["chr"],"align":["left"]},{"label":["lower"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["upper"],"name":[3],"type":["dbl"],"align":["right"]}],"data":[{"1":"Feces","2":"-0.80","3":"0.95"},{"1":"Freshwater","2":"NaN","3":"0.00"},{"1":"Freshwater (creek)","2":"NA","3":"0.87"},{"1":"Mock","2":"-0.87","3":"0.50"},{"1":"Ocean","2":"-1.00","3":"0.87"},{"1":"Sediment (estuary)","2":"NA","3":"0.87"},{"1":"Skin","2":"-1.00","3":"0.87"},{"1":"Soil","2":"NA","3":"0.00"},{"1":"Tongue","2":"NaN","3":"0.00"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<pre class="r"><code>quantile_permuted_rhos(permuted_rhos, p = 0.05, by_treatment = FALSE)</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["lower"],"name":[1],"type":["dbl"],"align":["right"]},{"label":["upper"],"name":[2],"type":["dbl"],"align":["right"]}],"data":[{"1":"-0.06","2":"-1"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<p><br></p>
<hr />
<p><br></p>
</div>
<div id="variable_correlation" class="section level2">
<h2>variable_correlation</h2>
<p><a name="variable_correlation"></a></p>
<p>Computes the correlation of numerical variables with taxa abundances. Function from the phylosmith-package.</p>
<p><br> <strong><em>Usage</em></strong></p>
<pre class="r"><code> variable_correlation(phyloseq_obj, variables, treatment = NULL,subset = NULL,
  classification = NULL, method = &#39;spearman&#39;, cores = 1)</code></pre>
<p><br> <strong><em>Arguments</em></strong></p>
<table>
<colgroup>
<col width="25%" />
<col width="75%" />
</colgroup>
<thead>
<tr class="header">
<th>Call</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>phyloseq_obj</code></td>
<td>A phyloseq-class object. It must contain <code>sample_data</code> with information about each sample, and it must contain <code>tax_table</code> with information about each taxa/gene.</td>
</tr>
<tr class="even">
<td><code>variables</code></td>
<td>Numerical factors within the in the sample_data to correlate with the abundance data.</td>
</tr>
<tr class="odd">
<td><code>treatment</code></td>
<td>Column name as a <code>string</code> or <code>numeric</code> in the <code>sample_data</code>. This can be a <code>vector</code> of multiple columns and they will be combined into a new column.</td>
</tr>
<tr class="even">
<td><code>subset</code></td>
<td>A factor within the <code>treatment</code>. This will remove any samples that to not contain this factor. This can be a vector of multiple factors to subset on.</td>
</tr>
<tr class="odd">
<td><code>classification</code></td>
<td>Column name as a string or numeric in the tax_table for the factor to conglomerate by.</td>
</tr>
<tr class="even">
<td><code>method</code></td>
<td>Which correlation method to calculate, “pearson”, “spearman”.</td>
</tr>
</tbody>
</table>
<p><code>cores</code> numeric value for number of CPU cores to use for the pair-wise permutations. Default (1), <code>0</code> uses max cores available. Parallelization not available for systems running MacOS without openMP configuration.</p>
<p><strong><em>Examples</em></strong></p>
<pre class="r"><code>data(enterotype)
filtered_obj &lt;- phyloseq::subset_samples(enterotype, 
  !is.na(enterotype@sam_data$ClinicalStatus))
filtered_obj &lt;- taxa_filter(filtered_obj, frequency = 0.65)

variable_correlation(filtered_obj, variables = &quot;Age&quot;,
 treatment = &quot;ClinicalStatus&quot;, subset = c(&quot;healthy&quot;, &quot;elderly&quot;, &quot;obese&quot;),
 classification = &quot;Genus&quot;, method = &quot;spearman&quot;)</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["Treatment"],"name":[1],"type":["chr"],"align":["left"]},{"label":["X"],"name":[2],"type":["chr"],"align":["left"]},{"label":["Y"],"name":[3],"type":["chr"],"align":["left"]},{"label":["rho"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["p"],"name":[5],"type":["dbl"],"align":["right"]}],"data":[{"1":"healthy","2":"Unclassified","3":"Age","4":"0.4108777397","5":"0.0514613831"},{"1":"healthy","2":"Oxalobacter","3":"Age","4":"0.5967771068","5":"0.0026465111"},{"1":"healthy","2":"Desulfovibrio","3":"Age","4":"0.5267127725","5":"0.0098183963"},{"1":"healthy","2":"Enterococcus","3":"Age","4":"-0.3199012005","5":"0.1367345208"},{"1":"healthy","2":"Escherichia/Shigella","3":"Age","4":"-0.4588381979","5":"0.0276478425"},{"1":"healthy","2":"Lactobacillus","3":"Age","4":"-0.2212323750","5":"0.3103600116"},{"1":"healthy","2":"Streptococcus","3":"Age","4":"-0.2091471527","5":"0.3381898166"},{"1":"healthy","2":"Clostridiales","3":"Age","4":"-0.2961681902","5":"0.1700070300"},{"1":"healthy","2":"Treponema","3":"Age","4":"0.1903957845","5":"0.3841992451"},{"1":"healthy","2":"Megasphaera","3":"Age","4":"0.1592482885","5":"0.4679680181"},{"1":"healthy","2":"Acidaminococcus","3":"Age","4":"0.0924598524","5":"0.6747816246"},{"1":"healthy","2":"Veillonella","3":"Age","4":"-0.2299135367","5":"0.2912617491"},{"1":"healthy","2":"Ruminococcaceae","3":"Age","4":"0.3535229650","5":"0.0979601039"},{"1":"healthy","2":"Subdoligranulum","3":"Age","4":"0.3332510187","5":"0.1202101841"},{"1":"healthy","2":"Faecalibacterium","3":"Age","4":"0.3144623856","5":"0.1439112851"},{"1":"healthy","2":"Acetivibrio","3":"Age","4":"0.1534273742","5":"0.4845934249"},{"1":"healthy","2":"Anaerotruncus","3":"Age","4":"0.1691394866","5":"0.4404031929"},{"1":"healthy","2":"Ruminococcus","3":"Age","4":"0.4390606894","5":"0.0360798572"},{"1":"healthy","2":"Clostridium","3":"Age","4":"-0.3115555739","5":"0.1478547620"},{"1":"healthy","2":"Peptostreptococcaceae","3":"Age","4":"-0.4257108711","5":"0.0428352710"},{"1":"healthy","2":"Sporacetigenium","3":"Age","4":"0.1933251459","5":"0.3767875907"},{"1":"healthy","2":"Peptostreptococcus","3":"Age","4":"0.1771838708","5":"0.4186381852"},{"1":"healthy","2":"Lachnospiraceae","3":"Age","4":"0.2892460622","5":"0.1806871550"},{"1":"healthy","2":"Oribacterium","3":"Age","4":"0.2007545219","5":"0.3583606867"},{"1":"healthy","2":"Coprococcus","3":"Age","4":"0.4039556117","5":"0.0559210587"},{"1":"healthy","2":"Roseburia","3":"Age","4":"0.4365885008","5":"0.0372627885"},{"1":"healthy","2":"Catonella","3":"Age","4":"0.2731106960","5":"0.2073494118"},{"1":"healthy","2":"Dorea","3":"Age","4":"0.4608159487","5":"0.0268997001"},{"1":"healthy","2":"Anaerostipes","3":"Age","4":"0.1354759334","5":"0.5376705875"},{"1":"healthy","2":"Eubacterium","3":"Age","4":"0.4672436390","5":"0.0245792988"},{"1":"healthy","2":"Blautia","3":"Age","4":"0.2086527150","5":"0.3393590465"},{"1":"healthy","2":"Marvinbryantia","3":"Age","4":"0.4604352690","5":"0.0270424314"},{"1":"healthy","2":"Catenibacterium","3":"Age","4":"0.1607319682","5":"0.4637777800"},{"1":"healthy","2":"Coprobacillus","3":"Age","4":"-0.0004944377","5":"0.9982135423"},{"1":"healthy","2":"Holdemania","3":"Age","4":"0.1211372397","5":"0.5819144561"},{"1":"healthy","2":"Bacteroides","3":"Age","4":"0.3436342107","5":"0.1083974083"},{"1":"healthy","2":"Parabacteroides","3":"Age","4":"0.4033655159","5":"0.0563145932"},{"1":"healthy","2":"Porphyromonas","3":"Age","4":"0.4224564289","5":"0.0446232881"},{"1":"healthy","2":"Prevotella","3":"Age","4":"0.5786350856","5":"0.0038206640"},{"1":"healthy","2":"Alistipes","3":"Age","4":"0.6691413221","5":"0.0004804142"},{"1":"healthy","2":"Actinomyces","3":"Age","4":"-0.2543924849","5":"0.2414450451"},{"1":"healthy","2":"Mobiluncus","3":"Age","4":"-0.1740703805","5":"0.4269915932"},{"1":"healthy","2":"Corynebacterium","3":"Age","4":"-0.1379096281","5":"0.5303199215"},{"1":"healthy","2":"Bifidobacterium","3":"Age","4":"-0.5033375921","5":"0.0143501744"},{"1":"healthy","2":"Collinsella","3":"Age","4":"0.2598366820","5":"0.2311734851"},{"1":"healthy","2":"Eggerthella","3":"Age","4":"-0.3298714548","5":"0.1242490867"},{"1":"healthy","2":"Gordonibacter","3":"Age","4":"0.4267664415","5":"0.0422675395"},{"1":"healthy","2":"Slackia","3":"Age","4":"0.3975029468","5":"0.0603412694"},{"1":"obese","2":"Unclassified","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Oxalobacter","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Desulfovibrio","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Enterococcus","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Escherichia/Shigella","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Lactobacillus","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Streptococcus","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Clostridiales","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Treponema","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Megasphaera","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Acidaminococcus","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Veillonella","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Ruminococcaceae","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Subdoligranulum","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Faecalibacterium","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Acetivibrio","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Anaerotruncus","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Ruminococcus","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Clostridium","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Peptostreptococcaceae","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Sporacetigenium","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Peptostreptococcus","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Lachnospiraceae","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Oribacterium","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Coprococcus","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Roseburia","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Catonella","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Dorea","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Anaerostipes","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Eubacterium","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Blautia","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Marvinbryantia","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Catenibacterium","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Coprobacillus","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Holdemania","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Bacteroides","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Parabacteroides","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Porphyromonas","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Prevotella","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Alistipes","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Actinomyces","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Mobiluncus","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Corynebacterium","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Bifidobacterium","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Collinsella","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Eggerthella","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Gordonibacter","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"obese","2":"Slackia","3":"Age","4":"0.0000000000","5":"1.0000000000"},{"1":"elderly","2":"Unclassified","3":"Age","4":"-0.6000000000","5":"0.2079999919"},{"1":"elderly","2":"Oxalobacter","3":"Age","4":"0.1428571429","5":"0.7871720117"},{"1":"elderly","2":"Desulfovibrio","3":"Age","4":"0.0285714286","5":"0.9571545190"},{"1":"elderly","2":"Enterococcus","3":"Age","4":"-0.4857142857","5":"0.3287230321"},{"1":"elderly","2":"Escherichia/Shigella","3":"Age","4":"0.2571428571","5":"0.6227871720"},{"1":"elderly","2":"Lactobacillus","3":"Age","4":"0.2000000000","5":"0.7040000000"},{"1":"elderly","2":"Streptococcus","3":"Age","4":"-0.5428571429","5":"0.2657026239"},{"1":"elderly","2":"Clostridiales","3":"Age","4":"-0.0285714286","5":"0.9571545190"},{"1":"elderly","2":"Treponema","3":"Age","4":"0.4285714286","5":"0.3965014577"},{"1":"elderly","2":"Megasphaera","3":"Age","4":"-0.3142857143","5":"0.5440932945"},{"1":"elderly","2":"Acidaminococcus","3":"Age","4":"0.0285714286","5":"0.9571545190"},{"1":"elderly","2":"Veillonella","3":"Age","4":"-0.5797710357","5":"0.2277839374"},{"1":"elderly","2":"Ruminococcaceae","3":"Age","4":"0.0857142857","5":"0.8717434402"},{"1":"elderly","2":"Subdoligranulum","3":"Age","4":"-0.2571428571","5":"0.6227871720"},{"1":"elderly","2":"Faecalibacterium","3":"Age","4":"0.1428571429","5":"0.7871720117"},{"1":"elderly","2":"Acetivibrio","3":"Age","4":"-0.4857142857","5":"0.3287230321"},{"1":"elderly","2":"Anaerotruncus","3":"Age","4":"0.4285714286","5":"0.3965014577"},{"1":"elderly","2":"Ruminococcus","3":"Age","4":"-0.0285714286","5":"0.9571545190"},{"1":"elderly","2":"Clostridium","3":"Age","4":"-0.4857142857","5":"0.3287230321"},{"1":"elderly","2":"Peptostreptococcaceae","3":"Age","4":"-0.4285714286","5":"0.3965014577"},{"1":"elderly","2":"Sporacetigenium","3":"Age","4":"-0.6000000000","5":"0.2079999919"},{"1":"elderly","2":"Peptostreptococcus","3":"Age","4":"-0.6571428571","5":"0.1561749265"},{"1":"elderly","2":"Lachnospiraceae","3":"Age","4":"-0.3142857143","5":"0.5440932945"},{"1":"elderly","2":"Oribacterium","3":"Age","4":"-0.5428571429","5":"0.2657026239"},{"1":"elderly","2":"Coprococcus","3":"Age","4":"-0.5428571429","5":"0.2657026239"},{"1":"elderly","2":"Roseburia","3":"Age","4":"-0.3142857143","5":"0.5440932945"},{"1":"elderly","2":"Catonella","3":"Age","4":"-0.6571428571","5":"0.1561749265"},{"1":"elderly","2":"Dorea","3":"Age","4":"-0.2000000000","5":"0.7040000000"},{"1":"elderly","2":"Anaerostipes","3":"Age","4":"-0.5428571429","5":"0.2657026239"},{"1":"elderly","2":"Eubacterium","3":"Age","4":"-0.0857142857","5":"0.8717434402"},{"1":"elderly","2":"Blautia","3":"Age","4":"-0.9428571429","5":"0.0048046647"},{"1":"elderly","2":"Marvinbryantia","3":"Age","4":"-0.6000000000","5":"0.2079999919"},{"1":"elderly","2":"Catenibacterium","3":"Age","4":"-0.9428571429","5":"0.0048046647"},{"1":"elderly","2":"Coprobacillus","3":"Age","4":"-0.4285714286","5":"0.3965014577"},{"1":"elderly","2":"Holdemania","3":"Age","4":"-0.0857142857","5":"0.8717434402"},{"1":"elderly","2":"Bacteroides","3":"Age","4":"-0.1428571429","5":"0.7871720117"},{"1":"elderly","2":"Parabacteroides","3":"Age","4":"-0.1428571429","5":"0.7871720117"},{"1":"elderly","2":"Porphyromonas","3":"Age","4":"-0.2000000000","5":"0.7040000000"},{"1":"elderly","2":"Prevotella","3":"Age","4":"-0.4857142857","5":"0.3287230321"},{"1":"elderly","2":"Alistipes","3":"Age","4":"0.3714285714","5":"0.4684781341"},{"1":"elderly","2":"Actinomyces","3":"Age","4":"-0.0857142857","5":"0.8717434402"},{"1":"elderly","2":"Mobiluncus","3":"Age","4":"-0.7775419144","5":"0.0687269414"},{"1":"elderly","2":"Corynebacterium","3":"Age","4":"0.6000000000","5":"0.2079999919"},{"1":"elderly","2":"Bifidobacterium","3":"Age","4":"0.6000000000","5":"0.2079999919"},{"1":"elderly","2":"Collinsella","3":"Age","4":"-0.5428571429","5":"0.2657026239"},{"1":"elderly","2":"Eggerthella","3":"Age","4":"0.5428571429","5":"0.2657026239"},{"1":"elderly","2":"Gordonibacter","3":"Age","4":"-0.0857142857","5":"0.8717434402"},{"1":"elderly","2":"Slackia","3":"Age","4":"-0.3142857143","5":"0.5440932945"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<p><br></p>
<hr />
</div>

<p><br>
<br>
<strong><a href="https://schuyler-smith.github.io/" target="_blank">Schuyler Smith</a></strong>
<br>
Ph.D. Bioinformatics and Computational Biology
<br>
</p>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = false;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
