<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Schuyler D. Smith" />


<title>phylosmith</title>

<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cerulean.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="src/styles.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
}

</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">phylosmith</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
    Home
  </a>
</li>
<li>
  <a href="wrangling.html">
    <span class="fa fa-file"></span>
     
    Wrangling
  </a>
</li>
<li>
  <a href="analytics.html">
    <span class="fa fa-chart-pie"></span>
     
    Analytics
  </a>
</li>
<li>
  <a href="graphics.html">
    <span class="fa fa-bar-chart"></span>
     
    Graphics
  </a>
</li>
<li>
  <a href="networks.html">
    <span class="fa fa-project-diagram"></span>
     
    Networks
  </a>
</li>
<li>
  <a href="calculations.html">
    <span class="fa fa-calculator"></span>
     
    Calculations
  </a>
</li>
<li>
  <a href="https://github.com/schuyler-smith/phylosmith/">
    <span class="fa fa-code"></span>
     
    Source Code
  </a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="http://github.com/schuyler-smith">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="http://germslab.org/">
    <span class="fa fa-flask"></span>
     
  </a>
</li>
<li>
  <a href="https://www.bcb.iastate.edu/people/schuyler-smith">
    <span class="fa fa-university"></span>
     
  </a>
</li>
<li>
  <a href="mailto:sdsmith@iastate.edu">
    <span class="fa fa-envelope fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">phylosmith</h1>
<h4 class="author"><a href="https://schuyler-smith.github.io/" target="_blank" >Schuyler D. Smith</a></h4>

</div>


<p><br></p>
<p><a href="https://github.com/schuyler-smith/phylosmith"><img src="https://travis-ci.org/schuyler-smith/phylosmith.svg?branch=master" alt="Travis Build Status" /></a> <a href="http://joss.theoj.org/papers/4d4780ab16c487764ebe108fa6bcdc2c"><img src="http://joss.theoj.org/papers/4d4780ab16c487764ebe108fa6bcdc2c/status.svg" alt="status" /></a> <a href="https://zenodo.org/badge/latestdoi/159598780"><img src="https://zenodo.org/badge/159598780.svg" alt="DOI" /></a></p>
<p><br></p>
<p><a href="https://github.com/schuyler-smith/phylosmith" targe="_blank" >phylosmith</a> is a supplementary package to build on the phyloseq-objecy from the <a href="https://github.com/joey711/phyloseq">phyloseq</a> package. Phyloseq-objects are a great data-standard for microbiome and gene-expression data, this package is aimed to provied easy data-wrangling and visualization.</p>
<p>A lot of these functions are just to make “data-wrangling” easier for investigators. Others will implement complex routines in a, hopefully, efficient and concise manner. I have also made functions to make figures for quick examination of data, but they may or may not be suitable for publication, as some may require parameter optimization.</p>
<p><br></p>
<div id="installation" class="section level1">
<h1>Installation</h1>
<div id="requirements" class="section level2">
<h2>Requirements</h2>
<div id="linux-systems" class="section level3">
<h3>Linux Systems</h3>
<p>For some Linux systems you may need to install the following two programs through your terminal.</p>
<p>Ubuntu example:</p>
<pre><code>sudo apt-get install libmysqlclient-dev libgdal-dev libudunits2-dev</code></pre>
<p>These programs are required by some dependancies and may not come in your default OS distribution.</p>
</div>
<div id="windows-systems" class="section level3">
<h3>Windows Systems</h3>
<p>if you are working on WINDOWS you likely need to install the CRAN program <a href="https://cran.r-project.org/bin/windows/Rtools/" target="_blank" >Rtools</a>.When prompted, select <code>add rtools to system PATH</code>.</p>
</div>
<div id="r" class="section level3">
<h3>R</h3>
<p>phylosmith depends on the usage of the <a href="https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0061217">phyloseq package</a> released by <a href="https://github.com/joey711">Dr. Paul McMurdie</a>. The pacakge is maintained on BioConductor, and can be installed through R using the following commands:</p>
<pre><code>if(!requireNamespace(&quot;BiocManager&quot;, quietly = TRUE)){
  install.packages(&quot;BiocManager&quot;)
} 
BiocManager::install(&quot;phyloseq&quot;)</code></pre>
<p>Additionally, the package imports a number of other packages to use their advanced functions. These packages may install with the phylosmith installation, but it is always best to install independently.</p>
<pre><code>install.packages(c(&quot;devtools&quot;, RcppEigen&quot;, &quot;RcppParallel&quot;, &quot;Rtsne&quot;, &quot;ggforce&quot;, &quot;units&quot;))</code></pre>
</div>
</div>
<div id="phylosmith" class="section level2">
<h2>phylosmith</h2>
<p>The package is hosted on Github, and can be installed through R with:</p>
<pre><code>devtools::install_github(&#39;schuyler-smith/phylosmith&#39;)
library(phylosmith)</code></pre>
<p><br></p>
</div>
</div>
<div id="functions" class="section level1">
<h1>Functions</h1>
<p><br></p>
<p><a name="wrangling"></a></p>
<div id="wrangling" class="section level3">
<h3>Wrangling</h3>
<table>
<colgroup>
<col width="25%" />
<col width="75%" />
</colgroup>
<thead>
<tr class="header">
<th>Call</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><a href="wrangling.html#conglomerate_samples">conglomerate_samples</a></td>
<td>combines samples based on common factor within sample_data</td>
</tr>
<tr class="even">
<td><a href="wrangling.html#conglomerate_taxa">conglomerate_taxa</a></td>
<td>combines taxa that have same classification</td>
</tr>
<tr class="odd">
<td><a href="wrangling.html#melt_phyloseq">melt_phyloseq</a></td>
<td>melts a phyloseq object into a data.table</td>
</tr>
<tr class="even">
<td><a href="wrangling.html#merge_treatments">merge_treatments</a></td>
<td>combines multiple columns in meta-data into a new column</td>
</tr>
<tr class="odd">
<td><a href="wrangling.html#set_sample_order">set_sample_order</a></td>
<td>sets the order of the samples of a phyloseq object</td>
</tr>
<tr class="even">
<td><a href="wrangling.html#set_treatment_levels">set_treatment_levels</a></td>
<td>sets the order of the factors in a sample_data column</td>
</tr>
<tr class="odd">
<td><a href="wrangling.html#taxa_extract">taxa_extract</a></td>
<td>creates a phyloseq-object containing only taxa of choice</td>
</tr>
<tr class="even">
<td><a href="wrangling.html#taxa_filter">taxa_filter</a></td>
<td>filter taxa by proportion of samples seen in</td>
</tr>
<tr class="odd">
<td><a href="wrangling.html#taxa_prune">taxa_prune</a></td>
<td>remove taxa from a phyloseq-object</td>
</tr>
</tbody>
</table>
<p><a name="analyitcs"></a></p>
</div>
<div id="analytics" class="section level3">
<h3>Analytics</h3>
<div id="normalization" class="section level4">
<h4>Normalization</h4>
<table>
<thead>
<tr class="header">
<th>Call</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><a href="analytics.html#library_size">library_size</a></td>
<td>transform abundance data using library-size normalization</td>
</tr>
<tr class="even">
<td><a href="analytics.html#relative_abundance">relative_abundance</a></td>
<td>transform abundance data to relative abundance</td>
</tr>
</tbody>
</table>
</div>
<div id="taxa" class="section level4">
<h4>Taxa</h4>
<table>
<thead>
<tr class="header">
<th>Call</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><a href="analytics.html#common_taxa">common_taxa</a></td>
<td>find taxa common to each treatment</td>
</tr>
<tr class="even">
<td><a href="analytics.html#taxa_core">taxa_core</a></td>
<td>filter taxa by proportion of samples and relative abundance</td>
</tr>
<tr class="odd">
<td><a href="analytics.html#taxa_proportions">taxa_proportions</a></td>
<td>computes the proportion of a taxa classification</td>
</tr>
<tr class="even">
<td><a href="analytics.html#unique_taxa">unique_taxa</a></td>
<td>find taxa unique to each treatment</td>
</tr>
</tbody>
</table>
<p><a name="graphics"></a></p>
</div>
</div>
<div id="graphics" class="section level3">
<h3>Graphics</h3>
<div id="abundance" class="section level4">
<h4>Abundance</h4>
<table>
<colgroup>
<col width="25%" />
<col width="75%" />
</colgroup>
<thead>
<tr class="header">
<th>Call</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><a href="graphics.html#abundance_heatmap">abundance_heatmap</a></td>
<td>create a ggplot object of the heatmaps of the abundance table</td>
</tr>
<tr class="even">
<td><a href="graphics.html#abundance_lines">abundance_lines</a></td>
<td>create a ggplot object of the abundance data as a line graph</td>
</tr>
<tr class="odd">
<td><a href="graphics.html#phylogeny_profile">phylogeny_profile</a></td>
<td>create a ggplot barplot object of the compositons of each sample at a taxonomic level</td>
</tr>
<tr class="even">
<td><a href="graphics.html#taxa_abundance_bars">taxa_abundance_bars</a></td>
<td>create a ggplot object of the abundance of taxa in each sample</td>
</tr>
<tr class="odd">
<td><a href="graphics.html#taxa_core_graph">taxa_core_graph</a></td>
<td>create a ggplot object of the core taxa over a range of parameters</td>
</tr>
<tr class="even">
<td><a href="graphics.html#variable_correlation_heatmap">variable_correlation_heatmap</a></td>
<td>create a ggplot heatmatp of the correlation of numerical variables with taxa</td>
</tr>
</tbody>
</table>
</div>
<div id="diversity" class="section level4">
<h4>Diversity</h4>
<table>
<colgroup>
<col width="25%" />
<col width="75%" />
</colgroup>
<thead>
<tr class="header">
<th>Call</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><a href="graphics.html#alpha_diversity_graph">alpha_diversity_graph</a></td>
<td>create a ggplot-object box-plot of the alpha-diversity from a phyloseq-object.</td>
</tr>
<tr class="even">
<td><a href="graphics.html#dendrogram_phyloseq">dendrogram_phyloseq</a></td>
<td>create a ggplot-object dendrogram of the distance measurement from a phyloseq-object.</td>
</tr>
<tr class="odd">
<td><a href="graphics.html#nmds_phyloseq">nmds_phyloseq</a></td>
<td>create a ggplot object of the NMDS from a phyloseq object</td>
</tr>
<tr class="even">
<td><a href="graphics.html#pcoa_phyloseq">pcoa_phyloseq</a></td>
<td>create a ggplot object of the PCoA from a phyloseq object</td>
</tr>
<tr class="odd">
<td><a href="graphics.html#tsne_phyloseq">tsne_phyloseq</a></td>
<td>create a ggplot object of the t-SNE from a phyloseq object</td>
</tr>
</tbody>
</table>
<p><a name="networks"></a></p>
</div>
</div>
<div id="networks" class="section level3">
<h3>Networks</h3>
<table>
<colgroup>
<col width="25%" />
<col width="75%" />
</colgroup>
<thead>
<tr class="header">
<th>Call</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><a href="networks.html#co_occurrence_network">co_occurrence_network</a></td>
<td>creates a network of the co-occurrence of taxa</td>
</tr>
<tr class="even">
<td><a href="networks.html#network_layout_ps">network_layout_ps</a></td>
<td>creates a layout object for a network</td>
</tr>
<tr class="odd">
<td><a href="networks.html#network_ps">network_ps</a></td>
<td>creates a network object from a phyloseq_object</td>
</tr>
<tr class="even">
<td><a href="networks.html#variable_correlation_network">variable_correlation_network</a></td>
<td>creates a network of the correlation of taxa and sample variables</td>
</tr>
</tbody>
</table>
<p><a name="calculations"></a></p>
</div>
<div id="calculations" class="section level3">
<h3>Calculations</h3>
<table>
<colgroup>
<col width="25%" />
<col width="75%" />
</colgroup>
<thead>
<tr class="header">
<th>Call</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><a href="calculations.html#co_occurrence">co_occurrence</a></td>
<td>calculate co-occurrence between taxa</td>
</tr>
<tr class="even">
<td><a href="calculations.html#curate_co_occurrence">curate_co_occurrence</a></td>
<td>filter co-occurrence tables</td>
</tr>
<tr class="odd">
<td><a href="calculations.html#permute_rho">permute_rho</a></td>
<td>runs permutations of the otu_table to calculate a significant <span class="math inline">\(\rho\)</span> value</td>
</tr>
<tr class="even">
<td><a href="calculations.html#histogram_permuted_rhos">histogram_permuted_rhos</a></td>
<td>Create a ggplot object of the distribution of rho values.</td>
</tr>
<tr class="odd">
<td><a href="calculations.html#quantile_permuted_rhos">quantile_permuted_rhos</a></td>
<td>calculate quantiles for the permuted rho values from the Spearman-rank co-occurrence</td>
</tr>
<tr class="even">
<td><a href="calculations.html#variable_correlation">variable_correlation</a></td>
<td>calculate the correlation of numerical variables with taxa abundances</td>
</tr>
</tbody>
</table>
</div>
<div id="datasets" class="section level3">
<h3>Datasets</h3>
<p>Originally I had created 2 mock phyloseq objects (<code>mock_phyloseq</code> and <code>mock_phyloseq2</code>) that had no real-world data but served to show simple examples of how the functions worked.</p>
<p>Then I decided that I should include a real example of microbiome data (<code>soil_column</code>) becasue it’s always nice to see real examples. <code>soil_column</code> is a <a href="https://www.frontiersin.org/articles/10.3389/fmicb.2018.03197/full" target="_blank" >published dataset</a> from my lab-group. The data is from an experiment where they looked at the microbial composition of farmland soil before and after manure application, over time, using 16S-sequencing.</p>
</div>
</div>

<p><br>
<br>
<strong><a href="https://schuyler-smith.github.io/" target="_blank">Schuyler Smith</a></strong>
<br>
Ph.D. Bioinformatics and Computational Biology
<br>
</p>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = false;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
