<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Schuyler D. Smith" />

<meta name="date" content="2023-03-26" />

<title>Networks</title>

<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cerulean.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="src/styles.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
}

</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">phylosmith</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
    Home
  </a>
</li>
<li>
  <a href="wrangling.html">
    <span class="fa fa-file"></span>
     
    Wrangling
  </a>
</li>
<li>
  <a href="analytics.html">
    <span class="fa fa-chart-pie"></span>
     
    Analytics
  </a>
</li>
<li>
  <a href="graphics.html">
    <span class="fa fa-bar-chart"></span>
     
    Graphics
  </a>
</li>
<li>
  <a href="networks.html">
    <span class="fa fa-project-diagram"></span>
     
    Networks
  </a>
</li>
<li>
  <a href="calculations.html">
    <span class="fa fa-calculator"></span>
     
    Calculations
  </a>
</li>
<li>
  <a href="https://github.com/schuyler-smith/phylosmith/">
    <span class="fa fa-code"></span>
     
    Source Code
  </a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="http://github.com/schuyler-smith">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="http://germslab.org/">
    <span class="fa fa-flask"></span>
     
  </a>
</li>
<li>
  <a href="https://www.bcb.iastate.edu/people/schuyler-smith">
    <span class="fa fa-university"></span>
     
  </a>
</li>
<li>
  <a href="mailto:sdsmith@iastate.edu">
    <span class="fa fa-envelope fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Networks</h1>
<h4 class="author"><a href="https://schuyler-smith.github.io/" target="_blank" >Schuyler D. Smith</a></h4>
<h4 class="date">2023-03-26</h4>

</div>


<p>Examples used in this vignette will use the <code>GlobalPatterns</code> dataset from <code>phyloseq</code>.</p>
<pre class="r"><code>library(phyloseq)
data(GlobalPatterns)</code></pre>
<p><br></p>
<div id="co_occurrence_network" class="section level2">
<h2>co_occurrence_network</h2>
<p><a name="co_occurrence_network"></a></p>
<p>Create a ggplot object of the co-occurrence network.</p>
<p><br> <strong><em>Usage</em></strong></p>
<pre class="r"><code>co_occurrence_network(phyloseq_obj, treatment = NULL, subset = NULL, co_occurrence_table = NULL, classification = NULL, node_colors = &#39;default&#39;, cluster = FALSE, cluster_colors = &#39;default&#39;, buffer = 0.5)</code></pre>
<p><br> <strong><em>Arguments</em></strong></p>
<table>
<colgroup>
<col width="25%" />
<col width="75%" />
</colgroup>
<thead>
<tr class="header">
<th>Call</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>phyloseq_obj</code></td>
<td>A phyloseq-class object.</td>
</tr>
<tr class="even">
<td><code>treatment</code></td>
<td>Column name as a <code>string</code>, or <code>vector</code> of, in the <code>sample_data</code>.</td>
</tr>
<tr class="odd">
<td><code>subset</code></td>
<td>A factor within the <code>treatment</code>. This will remove any samples that to</td>
</tr>
</tbody>
</table>
<p>not contain this factor. This can be a <code>vector</code> of multiple factors to subset on. <code>co_occurrence_table</code> | Table of the co-occurrence of taxa/genes in the phyloseq_obj, computed using <code>co_occurrence()</code>. If no table is given, it will be computed with the phyloseq_obj, using the given treatment and p = 0.05 with a rho cut-off of 0.8. <code>classification</code> | Column name as a <code>string</code> or <code>numeric</code> in the <code>tax_table</code> for the factor to use for node colors. <code>node_colors</code> | Name of a color set from the <a href="https://cran.r-project.org/web/packages/RColorBrewer/RColorBrewer.pdf" target="_blank" ><code>RColorBrewer</code></a> package or a vector palete of R-accepted colors. <code>cluster</code> | if <code>TRUE</code>, will use <code>igraph</code>’s <code>cluster_fast_greedy</code> method. Alternatively, you may pass a vector of cluster assignments with order corresponding to the order of the taxa_names in the <code>phyloseq_obj</code>. <code>cluster_colors</code> | Name of a color set from the <a href="https://cran.r-project.org/web/packages/RColorBrewer/RColorBrewer.pdf" target="_blank" ><code>RColorBrewer</code></a> package or a vector palete of R-accepted colors. <code>buffer</code> | Amount of space beyond the points to extend the cluster (aesthetic argument).</p>
<p><strong><em>Examples</em></strong> This sample set is too small to get any meaningful co-occurrence.</p>
<pre class="r"><code>filtered_obj &lt;- conglomerate_taxa(GlobalPatterns, &quot;Phylum&quot;)
co_occurrence_network(filtered_obj, treatment = &quot;SampleType&quot;, 
  classification = &#39;Phylum&#39;)</code></pre>
<p><img src="networks_files/figure-html/co_occurrence_network-1.png" width="1536" /></p>
<p><br></p>
<hr />
<p><br></p>
</div>
<div id="network_layout_ps" class="section level2">
<h2>network_layout_ps</h2>
<p><a name="network_layout_ps"></a></p>
<p>Create an layout_igraph object of the co-occurrence from a phyloseq object. This can be input into the co_occurrence_network function, or used for other network creating scripts. The purpose is to be able to create reproducible and comparable graphics.</p>
<p><br> <strong><em>Usage</em></strong></p>
<pre class="r"><code>network_layout_ps(phyloseq_obj, treatment = NULL, subset = NULL, co_occurrence_table = NULL, algorithm = &#39;fr&#39;)</code></pre>
<p><br> <strong><em>Arguments</em></strong></p>
<table>
<colgroup>
<col width="25%" />
<col width="75%" />
</colgroup>
<thead>
<tr class="header">
<th>Call</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>phyloseq_obj</code></td>
<td>A phyloseq-class object.</td>
</tr>
<tr class="even">
<td><code>treatment</code></td>
<td>Column name as a <code>string</code>, or <code>vector</code> of, in the <code>sample_data</code>.</td>
</tr>
<tr class="odd">
<td><code>subset</code></td>
<td>A factor within the <code>treatment</code>. This will remove any samples that to</td>
</tr>
</tbody>
</table>
<p>not contain this factor. This can be a <code>vector</code> of multiple factors to subset on. <code>co_occurrence_table</code> | Table of the co-occurrence of taxa/genes in the phyloseq_obj, computed using <code>co_occurrence()</code>. If no table is given, it will be computed with the phyloseq_obj, using the given treatment and p = 0.05 with a rho cut-off of 0.8. <code>algorithm</code> | Supported <code>igraph::layout_</code> algorithm.</p>
<p><strong><em>Examples</em></strong></p>
<pre class="r"><code>filtered_obj &lt;- conglomerate_taxa(GlobalPatterns, &quot;Phylum&quot;)
network_layout_ps(filtered_obj, treatment = &quot;SampleType&quot;, algorithm = &#39;kk&#39;)</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["x"],"name":[1],"type":["dbl"],"align":["right"]},{"label":["y"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["name"],"name":[3],"type":["chr"],"align":["left"]},{"label":["Kingdom"],"name":[4],"type":["chr"],"align":["left"]},{"label":["Phylum"],"name":[5],"type":["chr"],"align":["left"]},{"label":["Mean Relative Abundance"],"name":[6],"type":["chr"],"align":["left"]},{"label":[".ggraph.orig_index"],"name":[7],"type":["int"],"align":["right"]},{"label":["circular"],"name":[8],"type":["lgl"],"align":["right"]},{"label":[".ggraph.index"],"name":[9],"type":["int"],"align":["right"]}],"data":[{"1":"2.9112791","2":"-1.54907528","3":"Euryarchaeota","4":"Archaea","5":"Euryarchaeota","6":"(0,0.03]","7":"1","8":"FALSE","9":"1"},{"1":"-2.9253404","2":"1.95376686","3":"Actinobacteria","4":"Bacteria","5":"Actinobacteria","6":"(0.07,0.1]","7":"2","8":"FALSE","9":"2"},{"1":"-1.1721524","2":"-1.80061179","3":"MVP-15","4":"Bacteria","5":"MVP-15","6":"(0,0.03]","7":"3","8":"FALSE","9":"3"},{"1":"2.9568410","2":"-0.76202429","3":"SBR1093","4":"Bacteria","5":"SBR1093","6":"(0,0.03]","7":"4","8":"FALSE","9":"4"},{"1":"-3.1514314","2":"1.42525579","3":"Fusobacteria","4":"Bacteria","5":"Fusobacteria","6":"(0,0.03]","7":"5","8":"FALSE","9":"5"},{"1":"3.8313114","2":"4.52735856","3":"ZB3","4":"Bacteria","5":"ZB3","6":"(0,0.03]","7":"6","8":"FALSE","9":"6"},{"1":"0.9948009","2":"0.39605213","3":"GOUTA4","4":"Bacteria","5":"GOUTA4","6":"(0,0.03]","7":"7","8":"FALSE","9":"7"},{"1":"2.3268373","2":"6.39940588","3":"TG3","4":"Bacteria","5":"TG3","6":"(0,0.03]","7":"8","8":"FALSE","9":"8"},{"1":"4.6011230","2":"-1.55493618","3":"Chlorobi","4":"Bacteria","5":"Chlorobi","6":"(0,0.03]","7":"9","8":"FALSE","9":"9"},{"1":"-2.5164317","2":"2.36515916","3":"Bacteroidetes","4":"Bacteria","5":"Bacteroidetes","6":"(0.1,0.2]","7":"10","8":"FALSE","9":"10"},{"1":"-1.4348374","2":"-1.36151915","3":"KSB1","4":"Bacteria","5":"KSB1","6":"(0,0.03]","7":"11","8":"FALSE","9":"11"},{"1":"3.0582702","2":"1.18326058","3":"SAR406","4":"Bacteria","5":"SAR406","6":"(0,0.03]","7":"12","8":"FALSE","9":"12"},{"1":"0.4977171","2":"-1.84318514","3":"Thermi","4":"Bacteria","5":"Thermi","6":"(0,0.03]","7":"13","8":"FALSE","9":"13"},{"1":"1.3327685","2":"6.02675514","3":"Gemmatimonadetes","4":"Bacteria","5":"Gemmatimonadetes","6":"(0,0.03]","7":"14","8":"FALSE","9":"14"},{"1":"-0.9850066","2":"1.31968220","3":"Fibrobacteres","4":"Bacteria","5":"Fibrobacteres","6":"(0,0.03]","7":"15","8":"FALSE","9":"15"},{"1":"3.2560746","2":"5.79053615","3":"GN06","4":"Bacteria","5":"GN06","6":"(0,0.03]","7":"16","8":"FALSE","9":"16"},{"1":"1.8644890","2":"-3.60332379","3":"AC1","4":"Bacteria","5":"AC1","6":"(0,0.03]","7":"17","8":"FALSE","9":"17"},{"1":"-1.9580767","2":"-0.82111789","3":"TM6","4":"Bacteria","5":"TM6","6":"(0,0.03]","7":"18","8":"FALSE","9":"18"},{"1":"1.9379846","2":"-0.04637202","3":"OP8","4":"Bacteria","5":"OP8","6":"(0,0.03]","7":"19","8":"FALSE","9":"19"},{"1":"0.9420684","2":"2.35957372","3":"Elusimicrobia","4":"Bacteria","5":"Elusimicrobia","6":"(0,0.03]","7":"20","8":"FALSE","9":"20"},{"1":"-0.5030391","2":"-1.00643870","3":"NC10","4":"Bacteria","5":"NC10","6":"(0,0.03]","7":"21","8":"FALSE","9":"21"},{"1":"2.7540506","2":"5.16595013","3":"SPAM","4":"Bacteria","5":"SPAM","6":"(0,0.03]","7":"22","8":"FALSE","9":"22"},{"1":"-3.0313100","2":"0.25956842","3":"Unclassified Bacteria","4":"Bacteria","5":"Unclassified Bacteria","6":"(0,0.03]","7":"23","8":"FALSE","9":"23"},{"1":"1.6395592","2":"-1.92099119","3":"CCM11b","4":"Bacteria","5":"CCM11b","6":"(0,0.03]","7":"24","8":"FALSE","9":"24"},{"1":"-1.1432624","2":"0.87306494","3":"Nitrospirae","4":"Bacteria","5":"Nitrospirae","6":"(0,0.03]","7":"25","8":"FALSE","9":"25"},{"1":"-1.3205536","2":"-0.11840997","3":"NKB19","4":"Bacteria","5":"NKB19","6":"(0,0.03]","7":"26","8":"FALSE","9":"26"},{"1":"-0.3141877","2":"-1.70373529","3":"BRC1","4":"Bacteria","5":"BRC1","6":"(0,0.03]","7":"27","8":"FALSE","9":"27"},{"1":"-1.8632417","2":"-1.58393712","3":"Hyd24-12","4":"Bacteria","5":"Hyd24-12","6":"(0,0.03]","7":"28","8":"FALSE","9":"28"},{"1":"-3.0966705","2":"-0.35784953","3":"WS3","4":"Bacteria","5":"WS3","6":"(0,0.03]","7":"29","8":"FALSE","9":"29"},{"1":"0.1967377","2":"1.01478315","3":"GN04","4":"Bacteria","5":"GN04","6":"(0,0.03]","7":"30","8":"FALSE","9":"30"},{"1":"-1.7657769","2":"0.81197890","3":"Lentisphaerae","4":"Bacteria","5":"Lentisphaerae","6":"(0,0.03]","7":"31","8":"FALSE","9":"31"},{"1":"0.5262308","2":"-0.93718931","3":"LD1","4":"Bacteria","5":"LD1","6":"(0,0.03]","7":"32","8":"FALSE","9":"32"},{"1":"0.5734524","2":"-2.90843909","3":"Chlamydiae","4":"Bacteria","5":"Chlamydiae","6":"(0,0.03]","7":"33","8":"FALSE","9":"33"},{"1":"-0.5326639","2":"1.31877770","3":"OP3","4":"Bacteria","5":"OP3","6":"(0,0.03]","7":"34","8":"FALSE","9":"34"},{"1":"-1.9703646","2":"2.52846974","3":"Firmicutes","4":"Bacteria","5":"Firmicutes","6":"(0.1,0.2]","7":"35","8":"FALSE","9":"35"},{"1":"-1.0506801","2":"-0.24680498","3":"OP9","4":"Bacteria","5":"OP9","6":"(0,0.03]","7":"36","8":"FALSE","9":"36"},{"1":"3.8978802","2":"5.73722673","3":"WPS-2","4":"Bacteria","5":"WPS-2","6":"(0,0.03]","7":"37","8":"FALSE","9":"37"},{"1":"4.4901918","2":"4.91661496","3":"SC3","4":"Bacteria","5":"SC3","6":"(0,0.03]","7":"38","8":"FALSE","9":"38"},{"1":"-0.3929182","2":"0.56252985","3":"GN02","4":"Bacteria","5":"GN02","6":"(0,0.03]","7":"39","8":"FALSE","9":"39"},{"1":"2.6066015","2":"-3.14731197","3":"SM2F11","4":"Bacteria","5":"SM2F11","6":"(0,0.03]","7":"40","8":"FALSE","9":"40"},{"1":"4.6074350","2":"-0.27113148","3":"ZB2","4":"Bacteria","5":"ZB2","6":"(0,0.03]","7":"41","8":"FALSE","9":"41"},{"1":"-2.9816454","2":"-2.57675029","3":"OP11","4":"Bacteria","5":"OP11","6":"(0,0.03]","7":"42","8":"FALSE","9":"42"},{"1":"4.1845831","2":"-2.49822001","3":"Chloroflexi","4":"Bacteria","5":"Chloroflexi","6":"(0,0.03]","7":"43","8":"FALSE","9":"43"},{"1":"1.3094049","2":"-0.83683553","3":"SC4","4":"Bacteria","5":"SC4","6":"(0,0.03]","7":"44","8":"FALSE","9":"44"},{"1":"-0.6144391","2":"-0.23101296","3":"WS1","4":"Bacteria","5":"WS1","6":"(0,0.03]","7":"45","8":"FALSE","9":"45"},{"1":"-1.2311158","2":"-2.73655751","3":"GAL15","4":"Bacteria","5":"GAL15","6":"(0,0.03]","7":"46","8":"FALSE","9":"46"},{"1":"0.1109681","2":"2.70909505","3":"AD3","4":"Bacteria","5":"AD3","6":"(0,0.03]","7":"47","8":"FALSE","9":"47"},{"1":"0.7376388","2":"-0.21458336","3":"WS2","4":"Bacteria","5":"WS2","6":"(0,0.03]","7":"48","8":"FALSE","9":"48"},{"1":"3.4491698","2":"0.36378973","3":"Synergistetes","4":"Bacteria","5":"Synergistetes","6":"(0,0.03]","7":"49","8":"FALSE","9":"49"},{"1":"0.1398746","2":"0.15599864","3":"SR1","4":"Bacteria","5":"SR1","6":"(0,0.03]","7":"50","8":"FALSE","9":"50"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<p><br></p>
<hr />
<p><br></p>
</div>
<div id="network_ps" class="section level2">
<h2>network_ps</h2>
<p><a name="network_ps"></a></p>
<p>Create an layout_igraph object of the co-occurrence from a phyloseq object. This can be input into the co_occurrence_network function, or used for other network creating scripts. The purpose is to be able to create reproducible and comparable graphics.</p>
<p><br> <strong><em>Usage</em></strong> <code>rnetwork_ps(phyloseq_obj, treatment = NULL, subset = NULL, co_occurrence_table = NULL, rho = 0.6)</code></p>
<p><br> <strong><em>Arguments</em></strong></p>
<table>
<colgroup>
<col width="25%" />
<col width="75%" />
</colgroup>
<thead>
<tr class="header">
<th>Call</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>phyloseq_obj</code></td>
<td>A phyloseq-class object.</td>
</tr>
<tr class="even">
<td><code>treatment</code></td>
<td>Column name as a <code>string</code>, or <code>vector</code> of, in the <code>sample_data</code>.</td>
</tr>
<tr class="odd">
<td><code>subset</code></td>
<td>A factor within the <code>treatment</code>. This will remove any samples that to</td>
</tr>
</tbody>
</table>
<p>not contain this factor. This can be a <code>vector</code> of multiple factors to subset on. <code>co_occurrence_table</code> | Table of the co-occurrence of taxa/genes in the phyloseq_obj, computed using <code>co_occurrence()</code>. If no table is given, it will be computed with the phyloseq_obj, using the given treatment and p = 0.05 with a rho cut-off of 0.8. <code>rho</code> | Cutoffs to use to subset the <code>correlation_table</code> by correlation values.</p>
<p><strong><em>Examples</em></strong></p>
<pre class="r"><code>filtered_obj &lt;- conglomerate_taxa(GlobalPatterns, &quot;Phylum&quot;)
network_ps(filtered_obj, treatment = &quot;SampleType&quot;, rho = 0.6)</code></pre>
<pre><code>## IGRAPH 87d181c UN-- 50 165 -- 
## + attr: name (v/c), Kingdom (v/c), Phylum (v/c), Mean Relative
## | Abundance (v/c), Treatment (e/c), rho (e/n), p (e/n), Weight (e/n),
## | Edge (e/c)
## + edges from 87d181c (vertex names):
##  [1] SBR1093       --ZB2           SBR1093       --Synergistetes
##  [3] GOUTA4        --OP8           GOUTA4        --GN04         
##  [5] GOUTA4        --SR1           OP8           --GN04         
##  [7] OP8           --SR1           Hyd24-12      --OP11         
##  [9] GN04          --SR1           ZB2           --Synergistetes
## [11] Euryarchaeota --Chlorobi      Euryarchaeota --Chloroflexi  
## + ... omitted several edges</code></pre>
<p><br></p>
<hr />
<p><br></p>
</div>
<div id="variable_correlation_network" class="section level2">
<h2>variable_correlation_network</h2>
<p><a name="variable_correlation_network"></a></p>
<p>Create a network ggplot object of the correlation of taxa and sample variables from a phyloseq object. This may also be refered to as a bipartite graph. Function from the phylosmith-package.</p>
<p><br> <strong><em>Usage</em></strong></p>
<pre class="r"><code>variable_correlation_network(phyloseq_obj, variables, classification = NULL,
 treatment = NULL, subset = NULL, correlation_table = NULL, method = &#39;spearman&#39;,
 rho_threshold = c(-0.01, 0.01), p_threshold = 0.05, colors = &#39;default&#39;,
 negative_positive_colors = c(&#39;pink1&#39;, &#39;gray22&#39;))</code></pre>
<p><br> <strong><em>Arguments</em></strong></p>
<table>
<colgroup>
<col width="25%" />
<col width="75%" />
</colgroup>
<thead>
<tr class="header">
<th>Call</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>phyloseq_obj</code></td>
<td>A phyloseq-class object.</td>
</tr>
<tr class="even">
<td><code>co_occurrence_table</code></td>
<td>Table of the co-occurrence of taxa/genes in the phyloseq_obj, computed using <code>co_occurrence()</code>. If no table is given, it will be computed with the phyloseq_obj, using the given treatment and p = 0.05 with a rho cut-off of 0.8.</td>
</tr>
<tr class="odd">
<td><code>variables</code></td>
<td>Numerical factors within the in the sample_data to correlate with the abundance data.</td>
</tr>
<tr class="even">
<td><code>classification</code></td>
<td>Column name as a string or numeric in the tax_table for the factor to conglomerate by.</td>
</tr>
<tr class="odd">
<td><code>treatment</code></td>
<td>Column name as a <code>string</code>, or <code>vector</code> of, in the <code>sample_data</code>.</td>
</tr>
<tr class="even">
<td><code>subset</code></td>
<td>A factor within the <code>treatment</code>. This will remove any samples that to</td>
</tr>
</tbody>
</table>
<p>not contain this factor. This can be a <code>vector</code> of multiple factors to subset on. <code>correlation_table</code> | Table of the correlation of taxa/variables in the phyloseq_ob}, computed using <code>variable_correlation</code>. If no table is given, it will be computed with the phyloseq_obj, using the given treatment and p = 0.05. <code>method</code> | Which correlation method to calculate, “pearson”, “spearman”. <code>rho_threshold</code> | Cutoffs to use to subset the <code>correlation_table</code> by correlation values. <code>p_threshold</code> | Cutoffs to use to subset the <code>correlation_table</code> by singnificance values. <code>colors</code> | Name of a color set from the <a href="https://cran.r-project.org/web/packages/RColorBrewer/RColorBrewer.pdf" target="_blank" ><code>RColorBrewer</code></a> package or a vector palete of R-accepted colors. <code>negative_positive_colors</code> | colors to use for the edges to represent negative and positive correlations. Defaul (‘pink1’, ‘gray22’)</p>
<p><strong><em>Examples</em></strong></p>
<pre class="r"><code>filtered_obj &lt;- phyloseq::subset_samples(enterotype, 
  !is.na(enterotype@sam_data$ClinicalStatus))
filtered_obj &lt;- taxa_filter(filtered_obj, frequency = 0.65)
variable_correlation_network(filtered_obj, variables = &#39;Age&#39;,
 treatment = &quot;ClinicalStatus&quot;, subset = c(&quot;healthy&quot;, &quot;elderly&quot;, &quot;obese&quot;), 
 classification = &#39;Genus&#39;, method = &#39;spearman&#39;)</code></pre>
<p><img src="networks_files/figure-html/variable_correlation_network-1.png" width="1536" /></p>
<p><br></p>
<hr />
</div>

<p><br>
<br>
<strong><a href="https://schuyler-smith.github.io/" target="_blank">Schuyler Smith</a></strong>
<br>
Ph.D. Bioinformatics and Computational Biology
<br>
</p>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = false;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
